In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# 1. Graduate Admission Rate

Graduate admission rate: Multiple Linear Regression

In [12]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [58]:
#Import data
admission_rate=pd.read_csv('/Users/mirandidallas-fuge/Desktop/Admission_Predict.csv')

y=admission_rate['Chance of Admit ']
x=admission_rate.iloc[:,1:8]

In [13]:
# Fit the Model
reg = LinearRegression().fit(x, y)
# Print the Score
reg.score(x, y)
print("Multiple Linear Regression score:", reg.score(x, y))
print("Multiple Linear Regression MSE:", mean_squared_error(y, reg.predict(x)))
print("Multiple Linear Regression coef:", reg.coef_)

Multiple Linear Regression score: 0.8034713719824393
Multiple Linear Regression MSE: 0.003986893243246914
Multiple Linear Regression coef: [ 0.00173741  0.00291958  0.00571666 -0.00330517  0.02235313  0.11893945
  0.02452511]


Graduate admission rate: Lasso regression

In [16]:
from sklearn import linear_model
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler

In [17]:
# Scale data 
scaler = StandardScaler()
scaler.fit(x)
X = scaler.transform(x)

In [18]:
# Fit the Model and Print Score
lasso = linear_model.Lasso()
lasso.fit(X, y)
print("lasso score:", lasso.score(X, y))
print("lasso MSE:", mean_squared_error(y, lasso.predict(X)))
print("lasso coef:", lasso.coef_)

lasso score: 0.0
lasso MSE: 0.0202865775
lasso coef: [0. 0. 0. 0. 0. 0. 0.]


Graduate admission rate: Ridge regression

In [27]:
# Fit the Model and Print Score
ridge = linear_model.Ridge()
ridge.fit(X, y)

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [29]:
print("ridge score:",ridge.score(X, y))
print("ridge MSE:", mean_squared_error(y, ridge.predict(X)))
print("ridge coef:", ridge.coef_)

ridge score: 0.8034661029324273
ridge MSE: 0.003987000134238337
ridge coef: [ 0.02012161  0.01784985  0.00665018 -0.00311891  0.02005063  0.0701277
  0.01220766]


### Summary for graduate school data: 
- Multiple Linear Regression - MSE 0.003986893243246914
- Lasso Regression - MSE 0.0202865775
- Ridge Regression - MSE 0.003987000134238337

Thus since we want the smallest MSE of these, linear regression performs the best.

##### Data set attributes: 
- small dataset, low dimension, all numerical attributes

# 2. Crime incidence in community

In [366]:
#Import data
crime = pd.read_csv('/Users/mirandidallas-fuge/Desktop/CommViolPredUnnormalizedData.txt')

In [367]:
x = crime.iloc[:, 0:-1]
y = crime.iloc[:,-1]

X = pd.DataFrame(x)

In [368]:
df_0 = pd.get_dummies(X.iloc[:,0])
df_1 = pd.get_dummies(X.iloc[:,1])

In [369]:
x = x.drop(['BerkeleyHeightstownship', 'NJ'], axis=1)
df_new = pd.concat([x, df_0, df_1], axis=1)

In [370]:
df_new = df_new.replace('?', np.NaN)
imp = Imputer(missing_values='NaN', strategy='mean', axis=0)
imp.fit(df_new)
X_fixed= imp.transform(df_new)

Crime: multiple linear regression 

In [371]:
# Fit the Model
reg = LinearRegression().fit(X_fixed, y)
# Print the Score
reg.score(X_fixed, y)
print("Multiple Linear Regression score:", reg.score(X_fixed, y))
print("Multiple Linear Regression MSE:", mean_squared_error(y, reg.predict(X_fixed)))
print("Multiple Linear Regression coef:", reg.coef_)

ValueError: could not convert string to float: '?'

Crime: Lasso regression

In [ ]:
# Scale data 
scaler = StandardScaler()
scaler.fit(x)
X = scaler.transform(x)

In [ ]:
# Fit the Model and Print Score
lasso = linear_model.Lasso()
lasso.fit(X, y)
print("lasso score:", lasso.score(X, y))
print("lasso MSE:", mean_squared_error(y, lasso.predict(X)))
print("lasso coef:", lasso.coef_)

Crime: Ridge regression

In [ ]:
# Fit the Model and Print Score
ridge = linear_model.Ridge()
ridge.fit(X, y)

In [ ]:
print("ridge score:",ridge.score(X, y))
print("ridge MSE:", mean_squared_error(y, ridge.predict(X)))
print("ridge coef:", ridge.coef_)

### Summary for crime data: 
- Multiple Linear Regression - 
- Lasso Regression - 
- Ridge Regression - 

Thus since we want the smallest MSE of these, _________  performs the best.

##### Data set attributes: 
-

# 3. Automobile

Automobile: multiple linear regression 

In [372]:
import numpy as np
from sklearn.preprocessing import Imputer
cars=pd.read_csv('/Users/mirandidallas-fuge/Desktop/imports-85.data')

In [373]:
x = cars.iloc[:, 0:-1]
y = cars.iloc[:,-1]

In [374]:
X = pd.DataFrame(x)

In [375]:
df_2 = pd.get_dummies(X.iloc[:,2])
df_3 = pd.get_dummies(X.iloc[:,3])
df_4 = pd.get_dummies(X.iloc[:,4])
df_5 = pd.get_dummies(X.iloc[:,5])
df_6 = pd.get_dummies(X.iloc[:,6])
df_7 = pd.get_dummies(X.iloc[:,7])
df_8 = pd.get_dummies(X.iloc[:,8])
df_14 = pd.get_dummies(X.iloc[:,14])
df_15 = pd.get_dummies(X.iloc[:,15])
df_17 = pd.get_dummies(X.iloc[:,17])

In [376]:
x = x.drop(['alfa-romero', 'gas', 'std', 'two', 'convertible', 'rwd', 'front', 'dohc','four', 'mpfi'], axis=1)

In [377]:
df_new = pd.concat([x, df_2, df_3, df_4, df_5, df_6, df_7, df_8, df_14, df_15, df_17], axis=1)

In [378]:
df_new = df_new.replace('?', np.NaN)

In [379]:
imp = Imputer(missing_values='NaN', strategy='mean', axis=0)
imp.fit(df_new)
X_fixed= imp.transform(df_new)

In [380]:
# Fit the Model
reg = LinearRegression().fit(X_fixed, y)
# Print the Score
reg.score(X_fixed, y)
print("Multiple Linear Regression score:", reg.score(X_fixed, y))
print("Multiple Linear Regression MSE:", mean_squared_error(y, reg.predict(X_fixed)))
print("Multiple Linear Regression coef:", reg.coef_)

ValueError: could not convert string to float: '?'

Automobile: Lasso regression

In [334]:
# Scale data 
scaler = StandardScaler()
scaler.fit(x)
X = scaler.transform(x)

ValueError: could not convert string to float: '?'

In [ ]:
# Fit the Model and Print Score
lasso = linear_model.Lasso()
lasso.fit(X, y)
print("lasso score:", lasso.score(X, y))
print("lasso MSE:", mean_squared_error(y, lasso.predict(X)))
print("lasso coef:", lasso.coef_)

Automobile: Ridge regression

In [ ]:
# Fit the Model and Print Score
ridge = linear_model.Ridge()
ridge.fit(X, y)

In [ ]:
print("ridge score:",ridge.score(X, y))
print("ridge MSE:", mean_squared_error(y, ridge.predict(X)))
print("ridge coef:", ridge.coef_)

### Summary for automobile data: 
- Multiple Linear Regression - 
- Lasso Regression - 
- Ridge Regression - 

Thus since we want the smallest MSE of these, _________  performs the best.

##### Data set attributes: 
-

# 4. Breast Cancer 

Breast Cancer: Multiple Linear Regression 

In [217]:
Breast_cancer_wisconsin = pd.read_csv('/Users/mirandidallas-fuge/Desktop/breast-cancer-wisconsin.data')

In [228]:
x=Breast_cancer_wisconsin.iloc[:,2:]

In [229]:
y=Breast_cancer_wisconsin.iloc[:,1]

In [248]:
import pandas as pd
X = pd.DataFrame(x)

In [230]:
X = X.replace('?', np.NaN)

In [261]:
from sklearn.preprocessing import Imputer
imp = Imputer(missing_values='NaN', strategy='mean', axis=0)
imp.fit(X)
X_fixed= imp.transform(X)

In [262]:
# Fit the Model
reg = LinearRegression().fit(X_fixed, y)
# Print the Score
reg.score(X_fixed, y)
print("Multiple Linear Regression score:", reg.score(X_fixed, y))
print("Multiple Linear Regression MSE:", mean_squared_error(y, reg.predict(X_fixed)))
print("Multiple Linear Regression coef:", reg.coef_)

Multiple Linear Regression score: 0.5330725624458194
Multiple Linear Regression MSE: 3.701758257116035
Multiple Linear Regression coef: [ 0.08716186  0.16076712 -0.09082431 -0.01948014  0.00987729 -0.00819105
 -0.00554684  0.07202413  1.64568907]


Breast Cancer: Lasso

In [263]:
# Scale data 
scaler = StandardScaler()
scaler.fit(X_fixed)
X = scaler.transform(X_fixed)

In [264]:
# Fit the Model and Print Score
lasso = linear_model.Lasso()
lasso.fit(X_fixed, y)
print("lasso score:", lasso.score(X_fixed, y))
print("lasso MSE:", mean_squared_error(y, lasso.predict(X_fixed)))
print("lasso coef:", lasso.coef_)

lasso score: 0.4551149414498669
lasso MSE: 4.3197991859989155
lasso coef: [0.17375289 0.1993394  0.         0.         0.15732223 0.
 0.01514637 0.         0.        ]


Breast Cancer: Ridge Regression

In [265]:
# Fit the Model and Print Score
ridge = linear_model.Ridge()
ridge.fit(X_fixed, y)

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [266]:
print("ridge score:",ridge.score(X_fixed, y))
print("ridge MSE:", mean_squared_error(y, ridge.predict(X_fixed)))
print("ridge coef:", ridge.coef_)

ridge score: 0.5330684615436043
ridge MSE: 3.701790768695812
ridge coef: [ 0.0879629   0.16133473 -0.09066844 -0.01920239  0.01132331 -0.00753354
 -0.00499077  0.07213587  1.63168369]


### Summary for breast cancer data: 
- Multiple Linear Regression - 3.701758257116035
- Lasso Regression - 4.3197991859989155
- Ridge Regression - 3.701790768695812

Thus since we want the smallest MSE of these, multiple linear regression performs the best.

##### Data set attributes: 
- small dimensions

# 5. Parkinsons Data 

In [7]:
Parkinsons_data = pd.read_csv('/Users/mirandidallas-fuge/Desktop/parkinsons_updrs.data')

In [134]:
y = Parkinsons_data['total_UPDRS']

In [135]:
x = Parkinsons_data.drop(['total_UPDRS'], axis=1)

Parkinsons: Multiple Linear Regression 

In [136]:
# Fit the Model
reg = LinearRegression().fit(x, y)
# Print the Score
reg.score(x, y)
print("Multiple Linear Regression score:", reg.score(x, y))
print("Multiple Linear Regression MSE:", mean_squared_error(y, reg.predict(x)))
print("Multiple Linear Regression coef:", reg.coef_)

Multiple Linear Regression score: 0.9100144198179335
Multiple Linear Regression MSE: 10.30124083758564
Multiple Linear Regression coef: [ 4.72741397e-02  7.43271063e-02 -1.75884711e+00  2.64879855e-03
  1.20660278e+00 -3.56842204e+02  1.77854322e+04  8.27931046e+03
  8.22633190e+01 -2.60542785e+03 -5.08475929e+01  6.92595138e-01
 -1.36712611e+04  1.06861484e+02 -4.25065839e+01  4.55455275e+03
 -4.26252046e+00 -8.81464413e-02  2.85995358e+00 -3.94378918e+00
 -3.95847026e+00]


Parkinsons: Lasso Regression 

In [137]:
# Scale data 
scaler = StandardScaler()
scaler.fit(x)
X = scaler.transform(x)

In [138]:
# Fit the Model and Print Score
lasso = linear_model.Lasso()
lasso.fit(X, y)
print("lasso score:", lasso.score(X, y))
print("lasso MSE:", mean_squared_error(y, lasso.predict(X)))
print("lasso coef:", lasso.coef_)

lasso score: 0.8885117507280906
lasso MSE: 12.762792705087232
lasso coef: [ 0.          0.         -0.          0.          9.13478072  0.
  0.          0.         -0.          0.          0.          0.
  0.          0.          0.          0.         -0.         -0.
  0.         -0.          0.        ]


Parkinsons: Ridge Regression 

In [139]:
# Fit the Model and Print Score
ridge = linear_model.Ridge()
ridge.fit(X, y)

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [140]:
print("ridge score:",ridge.score(X, y))
print("ridge MSE:", mean_squared_error(y, ridge.predict(X)))
print("ridge coef:", ridge.coef_)

ridge score: 0.909998204912202
ridge MSE: 10.303097064425131
ridge coef: [ 0.58416848  0.6559799  -0.81844061  0.14182767  9.80558099 -1.96009871
  0.63471423  0.76440024  0.29214967  0.65528174 -1.27777731  0.1376577
 -0.07298332  1.76981864 -0.85086344 -0.02896885 -0.25614925 -0.38072316
  0.28682972 -0.28015241 -0.36377144]


### Summary for parkinson data: 
- Multiple Linear Regression - 10.30124083758564
- Lasso Regression - 12.762792705087232
- Ridge Regression - 10.303097064425131

Thus since we want the smallest MSE of these, linear performs the best but due to the large MSE values non of them perform that well.

##### Data set attributes: 
- small data set

# 6. Mercedes-Benz Greener Manufacturing

In [157]:
Mercedes = pd.read_csv('/Users/mirandidallas-fuge/Desktop/train_mercedes.csv')
y = Mercedes['y']
x = Mercedes.drop(['y'], axis=1)

fix categorical data

Mercedes-Benz Greener Manufacturing: Multiple Linear Regression

In [ ]:
# Fit the Model
reg = LinearRegression().fit(x, y)
# Print the Score
reg.score(x, y)
print("Multiple Linear Regression score:", reg.score(x, y))
print("Multiple Linear Regression MSE:", mean_squared_error(y, reg.predict(x)))
print("Multiple Linear Regression coef:", reg.coef_)

Mercedes-Benz Greener Manufacturing: Lasso Regression 

In [ ]:
# Scale data 
scaler = StandardScaler()
scaler.fit(x)
X = scaler.transform(x)

In [ ]:
# Fit the Model and Print Score
lasso = linear_model.Lasso()
lasso.fit(X, y)
print("lasso score:", lasso.score(X, y))
print("lasso MSE:", mean_squared_error(y, lasso.predict(X)))
print("lasso coef:", lasso.coef_)

Mercedes-Benz Greener Manufacturing: Ridge Regression 

In [ ]:
# Fit the Model and Print Score
ridge = linear_model.Ridge()
ridge.fit(X, y)

In [ ]:
print("ridge score:",ridge.score(X, y))
print("ridge MSE:", mean_squared_error(y, ridge.predict(X)))
print("ridge coef:", ridge.coef_)

### Summary for Mercedes data: 
- Multiple Linear Regression - 
- Lasso Regression - 
- Ridge Regression - 

Thus since we want the smallest MSE of these, _____ performs the best

##### Data set attributes: 
- 

# 7. Song Year Prediction 

In [351]:
song = pd.read_csv('/Users/mirandidallas-fuge/Desktop/YearPredictionMSD.txt')

In [359]:
y = song.iloc[:,0]
x = song.iloc[:,1:]

Song: Multiple Linear Regression 

In [360]:
# Fit the Model
reg = LinearRegression().fit(x, y)
# Print the Score
reg.score(x, y)
print("Multiple Linear Regression score:", reg.score(x, y))
print("Multiple Linear Regression MSE:", mean_squared_error(y, reg.predict(x)))
print("Multiple Linear Regression coef:", reg.coef_)

Multiple Linear Regression score: 0.23700077314993973
Multiple Linear Regression MSE: 91.1690694545467
Multiple Linear Regression coef: [ 8.75417828e-01 -5.63271810e-02 -4.36495594e-02  3.35266425e-03
 -1.47466963e-02 -2.20070428e-01 -6.73789497e-03 -1.00895620e-01
 -7.04729895e-02  2.50706569e-02 -1.65701122e-01 -1.85435368e-03
  4.70143401e-02  3.55103260e-04 -4.22581347e-04  5.99183918e-04
  4.76576350e-04  1.46662665e-03  1.92446773e-03  2.12835187e-03
  7.69875469e-04 -4.02598883e-04  7.53936242e-03  2.81199236e-03
 -3.55559424e-03  7.11398144e-05  1.58942049e-03  5.29413275e-04
  8.74563154e-04 -3.04158379e-04 -1.40496765e-03 -1.40135040e-03
 -5.55962706e-03  2.47233882e-03  1.84962263e-03 -5.29414513e-03
 -2.77269168e-04  6.79196361e-04  1.36514213e-03 -1.71044631e-03
 -1.99135793e-03 -7.64136852e-04 -1.40254468e-03 -2.35902050e-03
 -3.17994565e-03  6.81260459e-03  4.56073617e-04 -2.07499004e-03
  2.75161566e-04  1.94120580e-03  2.20109055e-04 -1.60484536e-03
  1.97086996e-03  4

Song: Lasso Regression

In [361]:
# Scale data 
scaler = StandardScaler()
scaler.fit(x)
X = scaler.transform(x)

In [362]:
# Fit the Model and Print Score
lasso = linear_model.Lasso()
lasso.fit(X, y)
print("lasso score:", lasso.score(X, y))
print("lasso MSE:", mean_squared_error(y, lasso.predict(X)))
print("lasso coef:", lasso.coef_)

lasso score: 0.09764067500655949
lasso MSE: 107.82089558978214
lasso coef: [ 1.68710504 -0.         -0.63795672  0.          0.         -0.86141556
  0.         -0.         -0.          0.          0.         -0.
  0.          0.29605591  0.          0.          0.          0.
  0.          0.52770142  0.          0.          0.          0.
  0.          0.          0.          0.          0.          0.
 -0.          0.         -0.          0.         -0.         -0.
 -0.          0.          0.         -0.00232873 -0.         -0.
 -0.         -0.         -0.         -0.          0.         -0.
 -0.         -0.          0.         -0.          0.         -0.
 -0.          0.         -0.          0.         -0.         -0.
 -0.         -0.         -0.         -0.         -0.         -0.
 -0.          0.         -0.          0.         -0.         -0.
  0.          0.          0.         -0.          0.         -0.
 -0.         -0.          0.         -0.          0.          0.
  0.   

Song: Ridge Regression

In [363]:
# Fit the Model and Print Score
ridge = linear_model.Ridge()
ridge.fit(X, y)

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [364]:
print("ridge score:",ridge.score(X, y))
print("ridge MSE:", mean_squared_error(y, ridge.predict(X)))
print("ridge coef:", ridge.coef_)

ridge score: 0.23700077314129467
ridge MSE: 91.16906945557967
ridge coef: [ 5.31160293e+00 -2.90534331e+00 -1.53943734e+00  5.46923690e-02
 -3.37113347e-01 -2.82957692e+00 -9.81646368e-02 -8.03506918e-01
 -7.45796459e-01  1.63720177e-01 -7.24224333e-01 -1.54414092e-02
  1.04651115e+00  6.21202788e-01 -5.33077829e-01  6.54810843e-01
  2.26711663e-01  8.46037751e-01  6.11013061e-01  6.58441345e-01
  1.64762301e-01 -6.67066205e-02  1.40954825e+00  4.31568378e-01
 -4.31745114e-01  5.09175936e-02  8.66034495e-01  1.15605873e-01
  1.42742202e-01 -4.10266604e-02 -1.39216946e-01 -1.00630539e-01
 -4.13699128e-01  1.32311792e-01  7.85886032e-02 -5.71301946e-01
 -1.15074828e-01  3.07948300e-01  3.55821050e-01 -3.51905638e-01
 -2.38634044e-01 -9.15784860e-02 -1.03259693e-01 -9.05238537e-02
 -1.32111584e-01  3.74513507e-01  2.14627667e-01 -5.44409368e-01
  5.76530964e-02  2.37123064e-01  2.05938188e-02 -1.20442028e-01
  1.37923342e-01  4.08665428e-02 -4.87585561e-03  4.45998036e-02
 -5.90419691e-01

### Summary for Song data: 
- Multiple Linear Regression -  91.1690694545467
- Lasso Regression - 107.82089558978214
- Ridge Regression - 91.16906945557967

Thus since we want the smallest MSE of these, none of these models perform well 

##### Data set attributes: 
- lots of columns

# 8. Solar Flares

In [185]:
solar = pd.read_csv('/Users/mirandidallas-fuge/Desktop/flare.data2', sep = " ")

In [199]:
x = solar.iloc[:,1:6]
y = solar.iloc[:,7]

Solar: Multiple Linear Regression

In [201]:
# Fit the Model
reg = LinearRegression().fit(x, y)
# Print the Score
reg.score(x, y)
print("Multiple Linear Regression score:", reg.score(x, y))
print("Multiple Linear Regression MSE:", mean_squared_error(y, reg.predict(x)))
print("Multiple Linear Regression coef:", reg.coef_)

Multiple Linear Regression score: 0.07079476730209577
Multiple Linear Regression MSE: 0.0851232049849271
Multiple Linear Regression coef: [0.06338076 0.02650041 0.02537617 0.42973347 0.        ]


Solar: Lasso Regression

In [202]:
# Scale data 
scaler = StandardScaler()
scaler.fit(x)
X = scaler.transform(x)

In [203]:
# Fit the Model and Print Score
lasso = linear_model.Lasso()
lasso.fit(X, y)
print("lasso score:", lasso.score(X, y))
print("lasso MSE:", mean_squared_error(y, lasso.predict(X)))
print("lasso coef:", lasso.coef_)

lasso score: 0.0
lasso MSE: 0.0916086156098969
lasso coef: [0. 0. 0. 0. 0.]


Solar: Ridge Regression 

In [204]:
# Fit the Model and Print Score
ridge = linear_model.Ridge()
ridge.fit(X, y)

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [205]:
print("ridge score:",ridge.score(X, y))
print("ridge MSE:", mean_squared_error(y, ridge.predict(X)))
print("ridge coef:", ridge.coef_)

ridge score: 0.07079472336923387
ridge MSE: 0.08512320900955576
ridge coef: [0.02029078 0.013006   0.00838167 0.06745725 0.        ]


### Summary for Solar data: 
- Multiple Linear Regression - 0.0851232049849271
- Lasso Regression - 0.0916086156098969
- Ridge Regression - 0.08512320900955576

Thus since we want the smallest MSE of these, multiple linear regression performs the best 

##### Data set attributes: 
- small data set